### PR TITLE
Specification for the LFP/EEG reports

### PR DESCRIPTION
**Pull request by joseph-tharayil**
As requested in [VIZTM-1150](https://bbpteam.epfl.ch/project/issues/browse/VIZTM-1150), I've added the specification for the LFP/EEG reports.

### CODE REVIEW

File: source/sonata_simulation.rst
Comment (jdcourcol) [inline source/sonata_simulation.rst:115]: we use population, node_ids in SONATA to identify cells. so this this to be adjusted.

File: source/sonata_simulation.rst
Comment (jdcourcol) [inline source/sonata_simulation.rst:110]: you can call it position then. It can be float32 as neuron are already in float32.
Also, please add the unit.

File: source/sonata_simulation.rst
Comment (jdcourcol) [inline source/sonata_simulation.rst:111]: you can put all your strings as utf8.

**Comment by jplanasc** [comment]
We'd like to review this PR before it's merged. @joseph-tharayil , @jamesgking , @jdcourcol can you please give us 2-4 weeks time? We're quite overloaded right now.
Thanks!

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:116]: `GIDs` isn't a SONATA concept; do you mean node_id?

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:110]: how does this differ from `region`? Not all regions have layers

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:113]: I think we should specify what `EEG` and `LFP` mean; ie how is the processing different

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:114]: since this is optional, what happens if it doesn't exist?

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:111]: why is this optional?

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:108]: `Field` is the name we use for other h5 files

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:100]: I think the file format could be added to the `sonata_tech.rst` file, instead of `SONATA Simulation Configuration file` page

**Review by mgeplf — COMMENTED** [review]
This is just a first pass; I need to think about the layout of the file more.

Could you add en example file?

File: source/sonata_simulation.rst
Comment (joseph-tharayil) [reply source/sonata_simulation.rst:100]: Is the intention of the `sonata_tech.rst` file to describe all of the input file formats? Since the electrodes file seems to me to be quite different from the node and edge files, I think it might be better to put the documentation in a dedicated page. That said, I can put it in `sonata_tech.rst` if that's the consensus opinion

File: source/sonata_simulation.rst
Comment (NadirRoGue) [inline source/sonata_simulation.rst:115]: In previous entries, the format is **/electrodes/{electrodename}**, then there is **/electrodes/{populationname}** .
The way I understand from this spec is that the same **electrodename** can affect different **populationname**, is that correct?
If that is the case, how does /electrode/{electrodename}/offset works, given that is a single value that could affect multiple populations with different node count?

**Comment by jplanasc** [comment]
@joseph-tharayil  I don't have access to that file, but in any case, I think the repository should be self-contained with the examples.

File: source/sonata_simulation.rst
Comment (mgeplf) [reply source/sonata_simulation.rst:100]: `sonata_tech.rst` covers "The following is the definition of the file format used in the SONATA file.", so I think it's an appropriate place.

File: source/sonata_simulation.rst
Comment (mgeplf) [reply source/sonata_simulation.rst:113]: I don't understand what the `{electrodename}` does; it appears to be metadata, but it's not clear where any of it is used. Maybe I'm missing something, but it doesn't seem to be referenced in the `/electrodes/{population_name}/*` or `/{population_name}/*` datasets

File: source/sonata_simulation.rst
Comment (mgeplf) [reply source/sonata_simulation.rst:114]: I'm not sure what it's meant to represent, and if it's not used, I'm not sure we should add it, since it distracts from understanding the format.

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:115]: What happens if an `{electrodename}` is the same as a `{population_name}`? Or, to phrase it a different way, how can one distinguish `{electrodename}` from `{population_name}` in the `/electrodes` namespace?

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:118]: I'm concerned about the layout of this portion: it appears that there are datasets for potentially millions of cells:

```
/nodeA                   Group
/nodeA/node_ids          Dataset {2}
/nodeA/offsets           Group
/nodeA/offsets/0         Dataset {330}
(...potentially millions of datasets...)
/nodeA/offsets/N         Dataset {245}

/nodeA/sec_ids           Group
/nodeA/sec_ids/0         Dataset {1644}
(...potentially millions of datasets...)
/nodeA/sec_ids/N         Dataset {1243}
```

My recollection is that having this order of magnitude is not great for performance, and can result in significant slowdowns.  I believe that HPC have looked at this, in the context of morphology containers: perhaps @pramodk / @1uc / @matz-e remember better than I do, and can comment.

File: source/sonata_simulation.rst
Comment (joseph-tharayil) [reply source/sonata_simulation.rst]: `layer` is only relevant for the cortex. As per the latest version of the spec, if the electrode is in a region without cortical layers, then `layer` should be"NA". If the electrode is outside the brain, then is should be "Outside"

File: source/sonata_simulation.rst
Comment (joseph-tharayil) [reply source/sonata_simulation.rst]: `electrodename` is used in the code that generates the weights. Due to limitations in HDF5, the file is first created with values of 0 for the scaling factors. When these values are populated, `electrodename/positions` is used. The other, optional fields are just metadata

File: source/sonata_simulation.rst
Comment (joseph-tharayil) [reply source/sonata_simulation.rst]: @jorblancoa are you ok with not including it?

Comment (jorblancoa) [reply source/sonata_simulation.rst]: yes, we shouldnt need it since all {electrodename} share the node_ids and sec_ids

File: source/sonata_simulation.rst
Comment (jorblancoa) [reply source/sonata_simulation.rst]: After conducting tests and discussing with @1uc  @matz-e  @joseph-tharayil , we observed that the creation of the HDF5 file slows down significantly after approximately 1 million datasets. This issue has been discussed in detail in the following GitHub threads: [h5py issue #1055](https://github.com/h5py/h5py/issues/1055) and [Hierarchical-Localization pull request #193](https://github.com/cvg/Hierarchical-Localization/pull/193).

Considering our situation with potentially several million node_ids, the previously mentioned solution may not be sustainable. As an alternative, I propose a structure similar to reports, utilizing the following layout:
```
/electrodes/nodeA/data    Dataset {total_sections, nelectrodes}
/nodeA/sec_ids            Dataset {total_sections}
/nodeA/node_ids           Dataset {n_node_ids}
/nodeA/offsets            Dataset {n_node_ids+1}
```
To obtain data from a node_id with index i, we can access the data dataset as follows:
`subset = electrodes_dataset[offsets_dataset[i]:offsets_dataset[i+1], :]`

I have personally tested this approach with approximately 5 million node_ids, and the performance for both writing and reading the file was satisfactory.

Let me know if you have any further questions or concerns.

Comment (mgeplf) [reply source/sonata_simulation.rst]: Can you update the sample file, please?

Comment (jorblancoa) [reply source/sonata_simulation.rst]: Done!

**Comment by jorblancoa** [comment]
Any other comments? Otherwise I would like to merge this @mgeplf @jdcourcol @jplanasc @NadirRoGue @joseph-tharayil

File: source/sonata_tech.rst
Comment (jdcourcol) [inline source/sonata_tech.rst:596]: unit64 -> uint64 (and below)

**Review by jdcourcol — DISMISSED** [review]
small typos. Otherwise, I think we can proceed.

File: source/sonata_tech.rst
Comment (jorblancoa) [reply source/sonata_tech.rst]: thanks! fixed!

File: source/sonata_tech.rst
Comment (mgeplf) [inline source/sonata_tech.rst:590]: What is the meaning of `contacts`? Everything under this dataset appears to be metadata, I think?

File: source/sonata_tech.rst
Comment (mgeplf) [inline source/sonata_tech.rst:594]: Can `data` have a more descriptive name, like `segment_scaling_factor`?

File: source/sonata_tech.rst
Comment (mgeplf) [inline source/sonata_tech.rst:596]: Can `sec_ids` have not be shortened? `section_ids` would more closely match how similar ones are named in the edges files.

File: source/sonata_tech.rst
Comment (mgeplf) [inline source/sonata_tech.rst:594]: How are multiple electrodes tied together, if they reside in the same population?  IE if we have
```
/electrodes/contacts/electrode0
/electrodes/contacts/electrode1
```
That both refer to `nodeA`, how does one know how they correspond to: `/electrodes/{population_name}` and `/{population_name}`?

Comment (jorblancoa) [reply source/sonata_tech.rst:594]: maybe 'scaling_factors' would be better? I think is descriptive enough, and 'segment_scaling_factor' I feel is too long

File: source/sonata_tech.rst
Comment (jorblancoa) [reply source/sonata_tech.rst:596]: i could change that

File: source/sonata_tech.rst
Comment (joseph-tharayil) [reply source/sonata_tech.rst:590]: Yes, it is all metadata

Comment (mgeplf) [reply source/sonata_tech.rst:590]: Why is it called `contacts`?

Comment (joseph-tharayil) [reply source/sonata_tech.rst]: Originally, the intent was to have each {electrodename} refer to a different contact of the same electrode array. Now that we allow recordings from multiple arrays, this has been changed to /electrodes/{array_name}/{electrodename}

File: source/sonata_tech.rst
Comment (joseph-tharayil) [reply source/sonata_tech.rst]: I've added an index field in the latest commit

**Comment by jorblancoa** [comment]
I have included a sample file based on the latest feedback. I believe it would be beneficial to merge this file to unblock the libsonata/neurodamus pull requests. Moving forward, we can create separate pull requests for any future improvements we want to make to the file format. 
@joseph-tharayil @jdcourcol @mgeplf

File: source/sonata_tech.rst
Comment (mgeplf) [inline source/sonata_tech.rst:596]: does this have to be all the node ids in the population?

File: source/sonata_tech.rst
Comment (mgeplf) [inline source/sonata_tech.rst:598]: `sec_ids` was renamed `section_ids`

File: source/sonata_tech.rst
Comment (mgeplf) [inline source/sonata_tech.rst:601]: `that gives the path of the circuit to be used.`; in the example file, it has `/path/to/nodes.h5`; this really impacts the portability of moving files around, since one would have to update the electrodes file each time a file is moved.  What is this required for?

**Review by mgeplf — APPROVED** [review]

File: source/sonata_tech.rst
Comment (jorblancoa) [reply source/sonata_tech.rst:596]: No, the ones not here will be ignored during the simulation. Basically the report will have the intersection.

File: source/sonata_tech.rst
Comment (jorblancoa) [reply source/sonata_tech.rst:601]: The idea of this was to avoid using an electrodes_file in a simulation which circuit doesn't match.
Right now we check if it matches otherwise we dont generate a report and throw a warning. It was suggested in this ticket
https://bbpteam.epfl.ch/project/issues/browse/BBPBGLIB-749

**Review by jorblancoa — APPROVED** [review]

File: source/sonata_tech.rst
Comment (mgeplf) [reply source/sonata_tech.rst:601]: Ah, I see.
What happens if the different electrodes require different node ids?

File: source/sonata_tech.rst
Comment (mgeplf) [reply source/sonata_tech.rst]: Ah, I see. What happens if the different electrodes require different node ids?

File: source/sonata_tech.rst
Comment (jorblancoa) [reply source/sonata_tech.rst:601]: I guess the same way the circuit_config needs to be updated, we could also update the electrodes_file associated with that circuit?

Comment (mgeplf) [reply source/sonata_tech.rst:601]: Normally people wouldn't need to change the `circuit_config.json`, since it works with relative paths.  Moreover, it's changing a text file, which is easy, changing attributes on an h5py is not as trivial.

Comment (mgeplf) [reply source/sonata_tech.rst:601]: The more I think about it, I'm not sure this path is needed - the `edges.h5` files don't refer to their `nodes.h5`, and the expectation is that if there is a dectable mismatch, then the simulator will raise an error (ie: if the morphology doesn't have enough sections); couldn't we do the same here?

Comment (jorblancoa) [reply source/sonata_tech.rst:601]: Its a couple of lines of code, but i get your point. Maybe we can think where to put the electrodes_file so we have relative paths instead, or see once we encounter the issue in the future what would be the best

Comment (jorblancoa) [reply source/sonata_tech.rst:601]: I think that is more clear and fast to fail when the circuit doesnt match instead of going through the sections/segments and see there is a missmatch. But probably @joseph-tharayil knows better here.

File: source/sonata_tech.rst
Comment (mgeplf) [inline source/sonata_tech.rst:595]: when describing `each segment`, is this the morphology segment (ie: between two sample points?), or the NEURON compartment?

If so, I'm not sure I follow why the section_ids field is needed, because (I think?) there is a 1-to-1 relationship between the segment and the section. Does that mean that `/{population_name}/section_ids` isn't needeed?

File: source/sonata_tech.rst
Comment (jdcourcol) [reply source/sonata_tech.rst:601]: indeed, we should not put a path (absolute or not) in the binary. Like other artefact, such as spike file, it is up to the user to make sure it is consistent.

File: source/sonata_tech.rst
Comment (jorblancoa) [reply source/sonata_tech.rst:595]: My understanding is that most of the times there is a 1-to-1 match, but if not, only the sections in 'section_ids' will be reported. Again @joseph-tharayil correct me if I'm wrong

File: source/sonata_tech.rst
Comment (jorblancoa) [reply source/sonata_tech.rst:601]: Maybe we should leave it as 'metadata'? even if we don't enforce/check during the simulation?
Or shall we remove it altogether?

Comment (jdcourcol) [reply source/sonata_tech.rst:601]: we don't have anything that identifies uniquely a circuit (like a checksum or uuid). So I would leave that out of that MR and remove it.

Comment (jorblancoa) [reply source/sonata_tech.rst]: done

**Review by jorblancoa — DISMISSED** [review]

**Comment by jorblancoa** [comment]
Any objections to merge this now @jdcourcol @mgeplf ?
For future request we can create a ticket under https://bbpteam.epfl.ch/project/issues/browse/BBPBGLIB-1033

**Comment by jdcourcol** [comment]
Can we refine/elaborate  the part related to segments and sections ? May be @joseph-tharayil can provide details.

File: source/sonata_tech.rst
Comment (joseph-tharayil) [reply source/sonata_tech.rst:595]: Segment does refer to the NEURON compartment, not the morphology segment (so there are several segments per section, with each segment being in only one section). I don't think there should ever be a case where we exclude sections from the recording, so I do think that `/{population_name}/section_ids` is not needed

Comment (jorblancoa) [reply source/sonata_tech.rst:595]: removed

**Comment by jorblancoa** [comment]
> Can we refine/elaborate the part related to segments and sections ? May be @joseph-tharayil can provide details.

I have removed it after Joseph comment

File: .DS_Store
Comment (mgeplf) [inline .DS_Store:1]: this file is not needed

File: source/sonata_simulation.rst
Comment (mgeplf) [inline source/sonata_simulation.rst:88]: Perhaps we should use the word `compartment` here (and else where `segments` are used)?

**Comment by jdcourcol** [comment]
I think you should then replace "segment" by "compartment" otherwise this is very misleading.

File: source/sonata_tech.rst
Comment (mgeplf) [reply source/sonata_tech.rst]: @jorblancoa / @joseph-tharayil ?

Comment (jorblancoa) [reply source/sonata_tech.rst]: We don't support different electrodes to report on different node ids

**Review by jorblancoa — APPROVED** [review]

File: source/sonata_tech.rst
Comment (mgeplf) [reply source/sonata_tech.rst]: Is there a physical/scientific reason for that?

This should probably describe then; something to the effect of `Electrode files are organized to have all the same NodeIds`

I would note that in the nodes and edges files, different populations are organized within the same file, and the same could be done here so that groups of electrodes with different node IDs can be supported by shifting things right; something like:
```
/electrodes/{electrode_group}/{electrodename}/[same as before]
/electrodes/{electrode_group}/{population_name}/node_ids
```

File: source/sonata_tech.rst
Comment (mgeplf) [inline source/sonata_tech.rst:594]: If the cardinality of the electrodes is low, is there a need for the indirection from this `index` to the column in `/electrodes/{population_name}/scaling_factors`; why not have a 1-d dataset under the `/electrodes/{electrodename}/{population_name}/scaling_factors`}

Comment (jorblancoa) [reply source/sonata_tech.rst:594]: could you give an example?

Comment (mgeplf) [reply source/sonata_tech.rst:594]: In `source/usecases/usecase1/electrodes_file/electrode_file.h5` currently:
```
HDF5 "source/usecases/usecase1/electrodes_file/electrode_file.h5" {
DATASET "/electrodes/Neuropixels-12_0/nodeA/index" {
   DATATYPE  H5T_STD_I64LE
   DATASPACE  SCALAR
   DATA {
   (0): 0
   }
}
```
Signifying using the column 0 of `/electrodes/nodeA/scaling_factors Dataset {2887, 12}`, correct?

Why not drop `/electrodes/nodeA/scaling_factors`, and directly put the 0th column under
`/electrodes/Neuropixels-12_0/nodeA/scaling_factors`?

Comment (jorblancoa) [reply source/sonata_tech.rst:594]: This would be inefficient to read since we want all the electrodes for each compartment (basically the row in the scaling_factors)

Comment (mgeplf) [reply source/sonata_tech.rst:594]: How many electrodes are generally being created?

Comment (mgeplf) [reply source/sonata_tech.rst:594]: Aren't the `Total_comp` pretty high (like O(1e9))? I'm surprised making one read vs. 384 would make that much difference, considering how much data will be coming across the line.

Comment (jorblancoa) [reply source/sonata_tech.rst:594]: Yes but in neurodamus you go gid by gid, it doesnt make sense to query all the electrode groups one by one when you can access the offsets for the gid and then get the subset from the scaling_factors directly

Comment (mgeplf) [reply source/sonata_tech.rst:594]: I'm not sure I follow; are you loading the data from the h5 all at once, or piece by piece on each query time?

I did a quick bench with:
```
import h5py
import numpy as np

def build(path, shape):
    data = np.random.random(shape)
    with h5py.File(path, 'w') as h5:
        h5.create_dataset('matrix', data=data)
        for i in range(shape[1]):
            h5.create_dataset(str(i), data=data[:, i])


def read_as_columns(path, shape):
    data = {}

    with h5py.File(path, 'r') as h5:
        for i in range(shape[1]):
            data[i] = h5[str(i)][:]

    return data

def read_as_matrix(path):
    with h5py.File(path, 'r') as h5:
        return h5['matrix'][:]

path = 'test.h5'
electrode_count = 400
shape = (1_000_000, electrode_count)

build(path, shape)
%timeit read_as_columns(path, shape)
%timeit read_as_matrix(path)
```
And I'm not seeing a large difference in loading time.

Comment (jorblancoa) [reply source/sonata_tech.rst:594]: i load all at once for each gid:

```
def get_node_id_subsets(self, node_id, population_name):
        node_ids = self._lfp_file[population_name]["node_ids"]
        # Look for the index of the node_id
        index = numpy.where(numpy.array(node_ids) == node_id)[0][0]
        offsets_dataset = self._lfp_file[population_name]["offsets"]
        electrodes_dataset = self._lfp_file["electrodes"][population_name]["scaling_factors"]
        index_low = offsets_dataset[index]
        index_high = offsets_dataset[index + 1]
        # Get the subsets for the data for the node_id
        subset_data = electrodes_dataset[index_low:index_high, :]
```

Comment (mgeplf) [reply source/sonata_tech.rst:594]: I see; I assume the plan is to move this all into libsonata once the standard is set, right?

Comment (jorblancoa) [reply source/sonata_tech.rst:594]: My idea was to have only the field for the electrodes_file (see here: https://github.com/BlueBrain/libsonata/pull/271) and that code to stay in neurodamus since the info about the population is there

Comment (mgeplf) [reply source/sonata_tech.rst:594]: > since the info about the population is there

Not sure what you mean; I thought that the viz team was also going to have to read this, sometimes, to make illustrations?

Comment (jorblancoa) [reply source/sonata_tech.rst:594]: But this would be for the metadata right? as I understand the data itself it will only be used by neurodamus to generate the LFP report

Comment (mgeplf) [reply source/sonata_tech.rst:594]: Not sure, generally things that are added to the spec get an implementation in libsonata.  In this case, I could see that it might be useful to be able to visualize the scaling factors on the actual morphologies.

As long as it doesn't impact the data layout, I don't think it matters, and we can cross that bridge in the future.

Comment (jorblancoa) [reply source/sonata_tech.rst:594]: I have created a ticket to implement the reading of the electrodes file in libsonata
https://bbpteam.epfl.ch/project/issues/browse/BBPBGLIB-1053

I think we should merge this and work on the libsonata implementation whenever we have resources available.
@jamesgking @pramodk @jdcourcol

**Review by mgeplf — APPROVED** [review]

**Comment by mgeplf** [comment]
I'm going to merge this; I still wonder if having electrods with different node_ids is a usecase:
https://github.com/BlueBrain/sonata-extension/pull/11#discussion_r1277475793

But I guess we can add that later, if necessary.

**Comment by jorblancoa** [comment]
> I'm going to merge this; I still wonder if having electrods with different node_ids is a usecase: [#11 (comment)](https://github.com/BlueBrain/sonata-extension/pull/11#discussion_r1277475793)
> 
> But I guess we can add that later, if necessary.

Thanks!
I've created a ticket for that
https://bbpteam.epfl.ch/project/issues/browse/BBPBGLIB-1054

**Pull request merged**
